src/annoylib.h

```diff
@@ -1172,7 +1206,87 @@ template<typename S, typename T, typename Distance, typename Random>
     return get_node_ptr<S, Node>(_nodes, _s, i);
   }
 
-  S _make_tree(const vector<S >& indices, bool is_root) {
+  void _thread_build(int q, int thread_idx, ThreadBarrier& barrier, std::mutex& _nodes_mutex) {
+    Random _random;
+    // Each thread needs its own seed, otherwise each thread would be building the same tree(s)
+    int seed = _is_seeded ? _seed + thread_idx : thread_idx;
+    _random.set_seed(seed);
+
+    vector<vector<Node*> > thread_trees;
+    vector<S> thread_roots;
+    while (1) {
+      if (q == -1) {
```

Comment:
this code to determine `q` if `q = -1` could probably just be left in the `build` function

Reply:
The thing with `q = -1`, is that I don't know beforehand how many trees each thread should build, because it depends on the number of nodes. That is why I have to check for `-1` everywhere.

A better solution would be to introduce a different meaning to `q = -1`, where each thread would build just one tree for example.

Reply:
Ah yeah good point. I forgot about that. Ok if your code works then let's keep it, I guess